tensorflow/core/kernels/training_ops_gpu.cu.cc

```diff
@@ -118,19 +118,172 @@ struct ApplyGradientDescent<GPUDevice, T> {
   }
 };
 
+#if TENSORFLOW_USE_ROCM
+
+// if any kernels involving complex sqrt/rsqrt are compiled with ROCm, build process completes without errors,
+// but the resulting executable ends up unusable (throwing errors "no device code available for function" 
+// for completely unrelated kernels.)
+// We have no choice but to implement sqrt and rsqrt by hand
+
+template <typename T>
+__device__ T impl_sqrt(T x) { return sqrt(x); }
+template <typename T>
+__device__ T impl_rsqrt(T x) { return rsqrt(x); }
+template <>
+__device__ Eigen::half impl_sqrt(Eigen::half x) { return __float2half(sqrt(__half2float(x))); }
+template <>
+__device__ Eigen::half impl_rsqrt(Eigen::half x) { return __float2half(rsqrt(__half2float(x))); }
```

Comment:
Can we try to add these functions on the eigen side of the fence? Am wondering whether doing so (or something equivalent) is all we need to avoid having to add `#ifdef TENSORFLOW_USE_ROCM` in all the `operator()` implementations

Don't quite know whether or not it will work, but it is worth a shot.  I can walk you through how to go about making the changes in Eigen. Ping me when you have the chance.

If the eigen experiment does not work, then we can proceed with the changes here.

Reply:
The downside of that approach is that the eigen pathway is less efficient. It does not fuse operations (I don't think) and something like this https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/pull/770/files#diff-74277ebeace05d36456297a10ebfe427R329-R331 compiles into ~4 separate GPU kernel launches instead of 1. We have too much framework overhead as is. 

We don't need #ifdef TENSORFLOW_USE_ROCM in every operator(), they are there to keep the CUDA pathway unaffected. We could drop them and that would enable the handcoded kernels for CUDA as well. But then we need to test that pathway too.

It would probably make more sense to get these functions into HIP headers instead of Eigen.

Reply:
agree with what you are saying.

you can merge this PR, and also file a PR to upstream this change. 

/cc @whchung